src/librustc_driver/lib.rs

```diff
@@ -202,10 +202,22 @@ pub fn run_compiler_with_file_loader<'a, L>(args: &[String],
     let codemap = Rc::new(CodeMap::with_file_loader(loader));
     let sess = session::build_session_with_codemap(sopts,
                                                    &dep_graph,
-                                                   input_file_path,
+                                                   input_file_path.clone(),
                                                    descriptions,
                                                    cstore.clone(),
                                                    codemap);
+
+    // This isn't used later in the compilation process, it's only used to check for overwrites of
+    // the source file.
+    let output_filenames = driver::build_output_filenames(&input, &odir, &ofile, &[], &sess);
+    match input_file_path {
+        Some(ifile) => if output_filenames.path(OutputType::Exe) == ifile {
```

Comment:
Hmm... Why checking only the executable output type? Rust can generate bytecode or assembly too and, as far as I know, all could overwrite the input file.